app/controllers/admin/reports_controller.rb

```diff
@@ -57,16 +57,16 @@ def render_data?
 
     def render_report_as(format)
       if OpenFoodNetwork::FeatureToggle.enabled?(:background_reports, spree_current_user)
-        blob = ReportJob.create_blob!
-        job = ReportJob.perform_later(
+        blob = ReportBlob.create_for_upload_later!
+        ReportJob.perform_later(
           report_class, spree_current_user, params, format, blob
         )
         Timeout.timeout(max_wait_time) do
-          sleep 1 until job.done?
+          sleep 1 until blob.content_stored?
```

Comment:
Nice extraction, but we now rely on the blob to know when the report is finished. That is, `content_stored?` could be true, but it doesn't know if more data will be written.

I'm not sure if it's possible, but I can imagine if this method call was made in the same instant that report data is being written, it could return only part of the report. 

Does the ActiveJob object have an API to say when it's completed? I can't find one..

Reply:
This is how we store it:
https://github.com/openfoodfoundation/openfoodnetwork/blob/3e89a110565950c84ee95553a15fc6a3598e3424/app/models/report_blob.rb#L23-L27

We upload it to the storage and then save the record in the database. The `content_stored?` is asking the database, not the actual storage. And the database is only updated once all data is there.

> Does the ActiveJob object have an API to say when it's completed?

Oh, I think that it's a blocking call. The docs: https://api.rubyonrails.org/v7.0.4.2/classes/ActiveStorage/Blob.html#method-i-upload

It mentions that it raises an exception if the checksum doesn't match when the content is uploaded. So it must wait for the whole content.